File: src/infrastructure/tcp/servers/game/commands/games/automatch-settings-block.ts
import { SETTINGS_BLOCK_SIZE } from "./automatch.ts";

// The 204-byte settings block carried inside 0x43f1 — the elected host hands
// it straight to the create-game task, whose 0x4310 builder reads nothing
// else. Every field here becomes the created game. Ported from the
// reference's AutomatchSettingsBlock; see its javadoc for the field evidence.
//
// Tiers: the observed retail values (DM 10/50, TDM 5/4/25, SNE 7/4) replace
// the client defaults for those rules; everything else stays at the client
// default — a known-wrong fixed value beats an invented one, and the
// difference is one line per rule as soon as a game of that rule is observed.

const ROTATION = 0;
const ROTATION_ENTRIES = 16;
const ENTRY_BYTES = 3; // rule, map, flags

const MAX_PLAYERS = 66;
const MAX_PLAYERS_VALUE = 16;
const CURRENT_PLAYERS = 67;
const PLAYERS_AT_CREATE = 1; // the host is the only player when it creates
const BRIEFING_TIME = 68;
const BRIEFING_SECONDS = 2;
const LEVEL_LIMIT_TOLERANCE = 95;
const LEVEL_CAP = 22;
const TIMERS = 100; // 17 u32s
const UNIQUE_BLUE = 169;
const COMMON_A = 177;
const COMMON_A_VALUE = 0x24;
const TEAM_KILL_KICK = 182;
const TEAM_KILL_KICK_VALUE = 3;
const SNAKE = 189;
const AUTOMATCH_SNAKE = 3;
const UNKNOWN_72 = 72;
const UNKNOWN_72_VALUE = 0x02000000;
const UNKNOWN_179 = 179;
const UNKNOWN_179_VALUE = 0x20;

// Timer-array index of each rule's first slot, and how many slots it owns
// (DM has no rounds slot — its count is fixed at one).
const RULE_TIMERS: Record<number, number[]> = {
  0: [9, 2], // Deathmatch      — time, tickets
  1: [6, 3], // Team Deathmatch — time, rounds, tickets
  2: [4, 2], // Rescue          — time, rounds
  3: [2, 2], // Capture         — time, rounds
  4: [0, 2], // Sneaking        — time, rounds (+ SNAKE)
  5: [11, 2], // Base           — time, rounds
  6: [13, 2], // Bomb           — time, rounds. Not served
  7: [15, 2], // Team Sneaking  — time, rounds. Not served
};

// Values a real automatch game was observed to use, by rule id.
const AUTOMATCH_TIMERS: Record<number, number[]> = {
  0: [10, 50], // DM:  10 min, 50 tickets          (default 5/30)
  1: [5, 4, 25], // TDM: 5 min, 4 rounds, 25 tickets (default 3/4/15)
  4: [7, 4], // SNE: 7 min, 4 rounds              (default 8/4)
};

// Client-default timer array: SNE 8/4, CAP 4/4, RES 4/4, TDM 3/4/15, DM 5/30,
// BASE 5/4, BOMB 30/4, TSNE 10/4.
const DEFAULT_TIMERS = [8, 4, 4, 4, 4, 4, 3, 4, 15, 5, 30, 5, 4, 30, 4, 10, 4];

function putU32(block: Uint8Array, at: number, value: number): void {
  const view = new DataView(block.buffer);
  view.setUint32(at, value >>> 0, false);
}

function putU16(block: Uint8Array, at: number, value: number): void {
  const view = new DataView(block.buffer);
  view.setUint16(at, value, false);
}

function template(): Uint8Array {
  const block = new Uint8Array(SETTINGS_BLOCK_SIZE);
  block[MAX_PLAYERS] = MAX_PLAYERS_VALUE;
  putU32(block, BRIEFING_TIME, BRIEFING_SECONDS);
  block[LEVEL_LIMIT_TOLERANCE] = LEVEL_CAP;
  for (let i = 0; i < DEFAULT_TIMERS.length; i++) {
    putU32(block, TIMERS + i * 4, DEFAULT_TIMERS[i]);
  }
  block[UNIQUE_BLUE] = 1;
  block[COMMON_A] = COMMON_A_VALUE;
  putU16(block, TEAM_KILL_KICK, TEAM_KILL_KICK_VALUE);
  block[SNAKE] = AUTOMATCH_SNAKE;
  // The two still-unexplained fields, emitted as captured.
  putU32(block, UNKNOWN_72, UNKNOWN_72_VALUE);
  block[UNKNOWN_179] = UNKNOWN_179_VALUE;
  return block;
}

/**
 * Builds the block for a match: one rotation entry per distinct rule the
 * group asked for, contiguous from index 0; flags left 0 (the one value the
 * disc's ruleopt_bit whitelist permits for every rule). Map must be nonzero —
 * the client discards a zero-map entry.
 */
export function buildAutomatchSettingsBlock(rules: number[], map: number): Uint8Array {
  if (rules.length === 0) throw new Error("An automatch rotation needs at least one rule.");
  if (rules.length > ROTATION_ENTRIES) {
    throw new Error(`At most ${ROTATION_ENTRIES} rotation entries, got ${rules.length}`);
  }
  if (map === 0) throw new Error("Automatch map must not be 0.");

  const block = template();
  block[CURRENT_PLAYERS] = PLAYERS_AT_CREATE;

  for (let entry = 0; entry < rules.length; entry++) {
    const at = ROTATION + entry * ENTRY_BYTES;
    block[at] = rules[entry];
    block[at + 1] = map;
    block[at + 2] = 0; // per-rule "rule option"
    applyObservedTimers(block, rules[entry]);
  }
  return block;
}

/** Replaces a rule's client defaults with the observed values where we have them. */
function applyObservedTimers(block: Uint8Array, rule: number): void {
  if (rule === 4) block[SNAKE] = AUTOMATCH_SNAKE;
  const observed = AUTOMATCH_TIMERS[rule];
  const slots = RULE_TIMERS[rule];
  if (!observed || !slots) return;
  for (let i = 0; i < observed.length && i < slots[1]; i++) {
    putU32(block, TIMERS + slots[0] * 4 + i * 4, observed[i]);
  }
}

File: src/GameLobbyServer/Commands/Game/Rooms/AutomatchSettingsBlock.cs
namespace Mgo2Server.GameLobbyServer.Commands.Game.Rooms;

/// <summary>
/// Builds the settings block an automatch hands to the elected host, which the
/// host passes straight to the room-creation task.
/// </summary>
public static class AutomatchSettingsBlock
{
    /// <summary>Size of the block.</summary>
    public const int Size = 204;

    /// <summary>Offset of the rotation.</summary>
    private const int Rotation = 0;

    /// <summary>Number of rotation entries.</summary>
    private const int RotationEntries = 16;

    /// <summary>Bytes one rotation entry occupies.</summary>
    private const int EntryBytes = 3;

    /// <summary>Offset of the player cap.</summary>
    private const int MaximumPlayers = 66;

    /// <summary>Player cap of an automatch room.</summary>
    private const int MaximumPlayersValue = 16;

    /// <summary>Offset of the current player count.</summary>
    private const int CurrentPlayers = 67;

    /// <summary>Players present when the host creates the room.</summary>
    private const int PlayersAtCreate = 1;

    /// <summary>Offset of the briefing time.</summary>
    private const int BriefingTime = 68;

    /// <summary>Briefing time of an automatch room.</summary>
    private const int BriefingSeconds = 2;

    /// <summary>Offset of the level-limit tolerance.</summary>
    private const int LevelLimitTolerance = 95;

    /// <summary>Level cap of an automatch room.</summary>
    private const int LevelCap = 22;

    /// <summary>Offset of the per-rule timer array.</summary>
    private const int Timers = 100;

    /// <summary>Offset of the unique-blue byte.</summary>
    private const int UniqueBlue = 169;

    /// <summary>Offset of the first common byte.</summary>
    private const int CommonA = 177;

    /// <summary>Value of the first common byte.</summary>
    private const int CommonAValue = 0x24;

    /// <summary>Offset of the team-kill kick limit.</summary>
    private const int TeamKillKick = 182;

    /// <summary>Value of the team-kill kick limit.</summary>
    private const int TeamKillKickValue = 3;

    /// <summary>Offset of the snake-defeat count.</summary>
    private const int Snake = 189;

    /// <summary>Snake-defeat count of an automatch room.</summary>
    private const int AutomatchSnake = 3;

    /// <summary>Offset of the first unexplained word.</summary>
    private const int UnknownWord = 72;

    /// <summary>Value of the first unexplained word.</summary>
    private const int UnknownWordValue = 0x02000000;

    /// <summary>Offset of the unexplained byte.</summary>
    private const int UnknownByte = 179;

    /// <summary>Value of the unexplained byte.</summary>
    private const int UnknownByteValue = 0x20;

    /// <summary>Timer slot and count of each rule's entries.</summary>
    private static readonly Dictionary<int, int[]> RuleTimers = new()
    {
        [0] = [9, 2],
        [1] = [6, 3],
        [2] = [4, 2],
        [3] = [2, 2],
        [4] = [0, 2],
        [5] = [11, 2],
        [6] = [13, 2],
        [7] = [15, 2],
    };

    /// <summary>Timer values a real automatch room was observed to use.</summary>
    private static readonly Dictionary<int, int[]> ObservedTimers = new()
    {
        [0] = [10, 50],
        [1] = [5, 4, 25],
        [4] = [7, 4],
    };

    /// <summary>Client-default timer array.</summary>
    private static readonly int[] DefaultTimers = [8, 4, 4, 4, 4, 4, 3, 4, 15, 5, 30, 5, 4, 30, 4, 10, 4];

    /// <summary>Builds the block for a match.</summary>
    /// <param name="rules">Rotation rules, entry zero first.</param>
    /// <param name="map">Map the match runs.</param>
    public static byte[] Build(IReadOnlyList<int> rules, int map)
    {
        if (rules.Count == 0)
        {
            throw new ArgumentException("An automatch rotation needs at least one rule.", nameof(rules));
        }

        if (rules.Count > RotationEntries)
        {
            throw new ArgumentException($"At most {RotationEntries} rotation entries are allowed.", nameof(rules));
        }

        if (map == 0)
        {
            throw new ArgumentException("An automatch map must not be zero.", nameof(map));
        }

        var block = Template();
        block[CurrentPlayers] = PlayersAtCreate;

        for (var entry = 0; entry < rules.Count; entry++)
        {
            var at = Rotation + entry * EntryBytes;
            block[at] = (byte)rules[entry];
            block[at + 1] = (byte)map;
            block[at + 2] = 0;
            ApplyObservedTimers(block, rules[entry]);
        }

        return block;
    }

    private static byte[] Template()
    {
        var block = new byte[Size];
        block[MaximumPlayers] = MaximumPlayersValue;
        WriteUInt32(block, BriefingTime, BriefingSeconds);
        block[LevelLimitTolerance] = LevelCap;

        for (var index = 0; index < DefaultTimers.Length; index++)
        {
            WriteUInt32(block, Timers + index * 4, DefaultTimers[index]);
        }

        block[UniqueBlue] = 1;
        block[CommonA] = CommonAValue;
        WriteUInt16(block, TeamKillKick, TeamKillKickValue);
        block[Snake] = AutomatchSnake;
        WriteUInt32(block, UnknownWord, UnknownWordValue);
        block[UnknownByte] = UnknownByteValue;
        return block;
    }

    private static void ApplyObservedTimers(byte[] block, int rule)
    {
        if (rule == 4)
        {
            block[Snake] = AutomatchSnake;
        }

        if (!ObservedTimers.TryGetValue(rule, out var observed) || !RuleTimers.TryGetValue(rule, out var slots))
        {
            return;
        }

        for (var index = 0; index < observed.Length && index < slots[1]; index++)
        {
            WriteUInt32(block, Timers + slots[0] * 4 + index * 4, observed[index]);
        }
    }

    private static void WriteUInt32(byte[] block, int at, int value) =>
        Mgo2Server.Shared.Utils.BinaryUtility.WriteUInt32BigEndian(block, at, (uint)value);

    private static void WriteUInt16(byte[] block, int at, int value) =>
        Mgo2Server.Shared.Utils.BinaryUtility.WriteUInt16BigEndian(block, at, (ushort)value);
}
